### PR TITLE
feat: Inhaltsverzeichnis für setup.md, customization.md und CONTRIBUTING.md

### LINKED ISSUE #405: feat: Inhaltsverzeichnis (ToC) für setup.md, customization.md und CONTRIBUTING.md
## Zusammenfassung

Alle drei großen Dokumentationsdateien (`docs/setup.md`, `docs/customization.md`, `CONTRIBUTING.md`) sollen ein automatisch generiertes Inhaltsverzeichnis erhalten – analog zum ToC in der README.md (#403 / #404).

## Motivation

| Datei | Zeilen | Überschriften (## + ###) | Git-☰ reicht? |
| ----- | -----: | -----------------------: | ------------- |
| `docs/setup.md` | ~370 | 23 | ❌ Lange Paket-Tabellen erschweren Navigation |
| `docs/customization.md` | ~300 | 22 | ❌ 7 Themenbereiche, Nutzer suchen gezielt |
| `CONTRIBUTING.md` | ~850 | ~59 | ❌ Komplexeste Datei, verschachtelt bis #### |

GitHubs eingebautes ☰-Icon ist erst nach Klick sichtbar. Ein explizites ToC gibt sofort Orientierung – besonders bei `CONTRIBUTING.md`, wo Entwickler zwischen Code-Konventionen, PR-Workflow und spezifischen Aufgaben navigieren.

Die README.md hat seit #404 ein dynamisches ToC. Die gleiche Logik sollte auf die restlichen Docs angewendet werden.

### 1. `heading_to_anchor()` und `generate_toc()` nach `common/` verschieben

Die beiden Funktionen aus `generators/readme.sh` (inkl. Slug-Deduplizierung und `LC_ALL=C.UTF-8`) in ein neues Modul `generators/common/toc.sh` extrahieren:

```zsh
# generators/common/toc.sh
heading_to_anchor() {
    local heading="$1"
    printf '%s' "$heading" \
        | LC_ALL=C.UTF-8 tr '[:upper:]' '[:lower:]' \
        | LC_ALL=C.UTF-8 sed 's/[^[:alnum:] -]//g; s/ /-/g'
}

generate_toc() {
    # ... bestehende Implementierung mit typeset -A slug_count ...
}
```

`common.sh` lädt das Modul automatisch, damit alle Generatoren darauf zugreifen können.

### 2. Generatoren erweitern

**`generators/setup.sh`** und **`generators/customization.sh`** auf die gleiche Zwei-Phasen-Architektur umstellen wie `readme.sh`:

1. Body in Variable generieren
2. `generate_toc()` auf Body anwenden
3. Header + ToC + Body zusammensetzen

**Position des ToC:** Direkt nach dem Seitentitel (erste H1-Überschrift).

### 3. CONTRIBUTING.md – Sonderbehandlung

`CONTRIBUTING.md` ist **manuell gepflegt** (kein Generator). Zwei Optionen:

**Option A (empfohlen):** Neuen Generator `generators/contributing.sh` erstellen, der das ToC generiert und per `generate-docs.sh --check` validiert. Der Rest des Inhalts bleibt manuell – nur das ToC ist generiert.

**Option B:** ToC manuell pflegen und einen Check im Pre-Commit-Hook integrieren, der Konsistenz prüft.

### 4. Tests

Für jede Datei:
- Unit-Tests für ToC-Einträge ≙ Überschriften aus Body (State-Machine-Ansatz, nicht globales `grep -c`)
- Edge Cases: Emojis in Überschriften, Unicode-Buchstaben, `####`-Überschriften ignoriert

### Betroffene Dateien

| Datei | Änderung |
| ----- | -------- |
| `generators/common/toc.sh` | Neues Modul mit `heading_to_anchor()` + `generate_toc()` |
| `generators/common.sh` | `source` für `toc.sh` ergänzen |
| `generators/readme.sh` | Lokale Funktionen entfernen, `common/toc.sh` nutzen |
| `generators/setup.sh` | Zwei-Phasen-Architektur + ToC-Integration |
| `generators/customization.sh` | Zwei-Phasen-Architektur + ToC-Integration |
| `docs/setup.md` | Regeneriert mit ToC |
| `docs/customization.md` | Regeneriert mit ToC |
| `CONTRIBUTING.md` | ToC-Sektion ergänzen (manuell oder generiert) |
| `tests/test-generator-setup.sh` | ToC-Konsistenz-Tests |
| `tests/test-generator-custom.sh` | ToC-Konsistenz-Tests |

## Bekannte Stolpersteine aus #404

Diese Fehler sind bereits in #404 aufgetreten und müssen vermieden werden:

| Fehler | Ursache | Lösung |
| ------ | ------- | ------ |
| **Locale-Bug** | `[:lower:]`/`[:upper:]` in C-Locale matcht kein Unicode | `LC_ALL=C.UTF-8` für `tr`/`sed` – **bereits in `heading_to_anchor()` gefixt** |
| **Fragiler ToC-Test** | Globales `grep -c '- ['` zählt auch Body-Links | State-Machine mit `in_toc`/`past_toc` Flags |
| **Irreführende Testnamen** | "Umlaute erhalten" testete keine Umlaute | Testdaten mit echten Umlauten (ü, ä, ö, ß) verwenden |
| **ZSH `local`-Bug** | `local var` in Schleife gibt vorherigen Wert aus | `local` einmal vor der Schleife deklarieren, in Schleife nur zuweisen |

## Alternativen

- **Kein ToC:** Auf GitHubs eingebautes ☰-Icon vertrauen. Nachteil: Erst nach Klick sichtbar, keine Sofort-Übersicht.
- **Nur setup.md + customization.md:** CONTRIBUTING.md hat die meisten Überschriften (~59), aber da manuell gepflegt, ist der Aufwand höher. Trotzdem empfohlen wegen der Komplexität.
- **Markdown-ToC-Tool (`doctoc`, `markdown-toc`):** Externe Abhängigkeit, nicht SSOT-konform, passt nicht in die Generator-Architektur.

### PR DESCRIPTION
**Pull request by tshofmann**
## Beschreibung

Erweitert alle drei großen Dokumentationsdateien um ein automatisch generiertes Inhaltsverzeichnis – analog zum ToC in der README.md (#404).

### Architektur

- `heading_to_anchor()` und `generate_toc()` aus `readme.sh` nach `generators/common/toc.sh` extrahiert
- Neue generische `inject_toc()`: teilt Content in Header/Body, generiert ToC aus Body, setzt alles zusammen
- `setup.sh` und `customization.sh` nutzen jetzt die Zwei-Phasen-Architektur (`_generate_*_content()` → `inject_toc()`)
- Neuer `contributing.sh` Generator: liest manuell gepflegtes CONTRIBUTING.md, ersetzt/ergänzt nur das ToC (eigene 3-Phasen State-Machine für Idempotenz)
- Code-Block-Tracking (`` ``` ``) in allen Parsern gegen false positives
- Bonus: `generate_toc()` aus `common/toc.sh` hat jetzt Code-Block-Tracking, das die alte lokale Version in `readme.sh` nicht hatte
- `assert_toc_consistency()` in `assertions.sh`: gemeinsamer Test-Helper für ToC-Vollständigkeit über alle 4 Generatoren (DRY)

### Betroffene Dateien

| Datei | Änderung |
| ----- | -------- |
| `generators/common/toc.sh` | **NEU** – `heading_to_anchor()`, `generate_toc()`, `inject_toc()` |
| `generators/contributing.sh` | **NEU** – ToC-Generator für CONTRIBUTING.md |
| `generators/common.sh` | `source toc.sh` ergänzt |
| `generators/readme.sh` | Lokale Funktionen entfernt (-55 Zeilen) |
| `generators/setup.sh` | Zwei-Phasen-Architektur + `inject_toc()` |
| `generators/customization.sh` | Zwei-Phasen-Architektur + `inject_toc()` |
| `generate-docs.sh` | CONTRIBUTING.md in `--check`/`--generate`/`--help` |
| `tests/lib/assertions.sh` | `assert_toc_consistency()` – gemeinsamer ToC-Test-Helper |
| `tests/test-generator-setup.sh` | ToC-Konsistenz via `assert_toc_consistency()` |
| `tests/test-generator-custom.sh` | ToC-Konsistenz via `assert_toc_consistency()` |
| `tests/test-generator-readme.sh` | ToC-Konsistenz via `assert_toc_consistency()` |
| `tests/test-generator-contributing.sh` | **NEU** – 16 Tests (Integration, Idempotenz, Code-Block, ToC-Ersetzung) |
| `docs/setup.md` | ToC eingefügt |
| `docs/customization.md` | ToC eingefügt |
| `CONTRIBUTING.md` | ToC eingefügt + "Was wird generiert?"-Tabelle um CONTRIBUTING.md ergänzt |

## Art der Änderung

- [x] ✨ Neues Feature
- [x] 📝 Dokumentation
- [x] ♻️ Refactoring
- [x] 🧪 Testing

## Checkliste

- [x] Ich habe die [Contributing Guidelines](CONTRIBUTING.md) gelesen
- [x] `./.github/scripts/generate-docs.sh --check` ist erfolgreich
- [x] `./.github/scripts/health-check.sh` zeigt keine Fehler
- [x] Neue Aliase/Funktionen haben Beschreibungskommentare (falls zutreffend)
- [x] Bei neuen Tools: Guard-Check vorhanden (falls zutreffend)
- [x] Screenshots in `docs/assets/` noch aktuell (falls zutreffend)

## Zusammenhängende Issues

Closes #405

## Validierung

- 144 Generator-Tests bestanden (73 readme + 16 setup + 39 customization + 16 contributing)
- 608 Tests gesamt (15 Test-Dateien)
- Idempotenz verifiziert (generate → check → generate → alle unverändert)
- Code-Block-Tracking gegen `## ` false positives in Markdown-Code-Blöcken
- `## Inhalt` exakter Match (kein Falsch-Positiv bei `## Inhaltsverzeichnis`)

### CODE REVIEW

File: .github/scripts/generators/common/toc.sh [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    local toc
+    toc=$(generate_toc "$body")
+
+    printf '%s' "$header"
+    echo "## Inhalt"
+    echo ""
+    printf '%s\n' "$toc"
+    echo ""
+    printf '%s\n' "$body"
```
[the remark this answers is not in the document]
Comment (tshofmann) [reply .github/scripts/generators/common/toc.sh:114]: Berechtigter Hinweis. Praktisch getestet: `inject_toc()` auf bereits ToC-haltigen Content erzeugt tatsächlich ein doppeltes `## Inhalt`.

**Einschätzung:** In der Praxis tritt das nicht auf – `inject_toc()` wird nur von `_generate_setup_content()` und `_generate_customization_content()` aufgerufen, die garantiert ToC-freien Raw-Content erzeugen. `contributing.sh` nutzt bewusst eine eigene State-Machine mit `skip_toc`-Phase (idempotent getestet).

`inject_toc()` idempotent zu machen würde unnötige Komplexität einführen (doppelte Skip-Logik, die contributing.sh bereits eigenständig löst). Stattdessen als **Vorbedingung dokumentiert** in abd4ba1:

> VORBEDINGUNG: content darf kein "## Inhalt" enthalten – die Funktion prüft/entfernt kein bestehendes ToC. Für Dateien mit vorhandenem ToC (z.B. CONTRIBUTING.md) stattdessen eine eigene State-Machine nutzen (siehe contributing.sh).

File: .github/scripts/generators/contributing.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+# Haupt-Generator für CONTRIBUTING.md
+# ------------------------------------------------------------
+# Liest CONTRIBUTING.md, trennt Header vom Body, überspringt ein
+# vorhandenes "## Inhalt"-Abschnitt, generiert ToC aus dem Body,
+# und setzt alles zusammen: Header + "## Inhalt" + ToC + Body.
+generate_contributing_md() {
+    local source_file="$DOTFILES_DIR/CONTRIBUTING.md"
+    [[ -f "$source_file" ]] || { err "CONTRIBUTING.md nicht gefunden"; return 1; }
+
+    local header="" body="" phase="header" in_code_block=false line
```
[the remark this answers is not in the document]
Comment (tshofmann) [reply .github/scripts/generators/contributing.sh:22]: Berechtigt – jetzt ergänzt in abd4ba1:

`test-generator-contributing.sh` mit **16 Tests**:
- **Integration**: ToC-Überschrift + bekannte Einträge (Quick Setup, Code-Konventionen, Häufige Aufgaben, Dokumentation)
- **ToC-Konsistenz**: State-Machine zählt ToC-Einträge vs. Body-Überschriften
- **Idempotenz**: Zweimaliger Durchlauf produziert identischen Output
- **Code-Block-Robustheit**: `## ` innerhalb von ``` wird korrekt ignoriert (Fixture-Test)
- **Fehlende Datei**: Exit-Code 1, keine Ausgabe
- **ToC-Ersetzung**: Bestehender ToC wird durch neuen ersetzt, kein Doppel-`## Inhalt`

Alle 15 Test-Dateien bestehen (inkl. der neuen).

File: .github/scripts/generators/common/toc.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+# 1. Kleinschreibung
+# 2. Nicht-erlaubte Zeichen entfernen (behält Buchstaben, Ziffern, Leerzeichen, Bindestriche)
+# 3. Leerzeichen → Bindestriche
+# Subshell mit LC_ALL=C.UTF-8: GNU tr (Ubuntu) konvertiert Unicode-Case nicht,
+# daher ZSH-eigene ${(L)} Expansion für Lowercase + sed für Zeichenfilterung.
+heading_to_anchor() {
+    (
+        LC_ALL=C.UTF-8
+        local heading="$1"
+        printf '%s' "${(L)heading}" | sed 's/[^[:alnum:] -]//g; s/ /-/g'
```
[the remark this answers is not in the document]
Comment (tshofmann) [reply .github/scripts/generators/common/toc.sh:23]: **Nicht zutreffend – der Vorschlag würde eine Regression einführen.**

### Warum die aktuelle Implementierung korrekt ist

1. **`LC_ALL` ist in ZSH `scalar-export-special`** – Locale-Variablen werden automatisch in die Process-Environment übernommen, auch ohne explizites `export`. Getestet:
   ```
   LC_ALL type: scalar-export-special
   → Reassign in Subshell → sed sieht C.UTF-8 ✓
   ```

2. **`LC_ALL` muss im Subshell-Scope stehen**, weil **beide** Operationen die UTF-8-Locale brauchen:
   - `${(L)heading}` – ZSH-Lowercase-Expansion für Unicode (Ü→ü, Ä→ä, Ö→ö)
   - `sed 's/[^[:alnum:] -]//g'` – Unicode-Zeichen als alphanumerisch erkennen

### Warum der Vorschlag scheitert

Der Vorschlag verschiebt `LC_ALL=C.UTF-8` NUR vor `sed`:
```zsh
printf '%s' "${(L)heading}" | LC_ALL=C.UTF-8 sed '...'
```

Das bricht `${(L)heading}`: In C-Locale bleibt `Ü` groß → Anker wird `Über-die-konfiguration` statt `über-die-konfiguration`. **GitHub-Anker sind lowercase** → kaputte Links.

### Beweis (LC_ALL=C LANG=C Umgebung)

| Variante | `"Über die Konfiguration"` | Korrekt? |
|----------|---------------------------|----------|
| **Aktuell** (LC_ALL im Subshell-Scope) | `über-die-konfiguration` | ✅ |
| **Vorschlag** (LC_ALL nur sed-Prefix) | `Über-die-konfiguration` | ❌ |
| Ohne LC_ALL | `ber-die-konfiguration` | ❌ |
